src/PackageVersions/FallbackVersions.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace PackageVersions;
+
+/**
+ * @internal
+ *
+ * This is a fallback for {@see \PackageVersions\Versions::getVersion()}
+ * Do not use this class directly: it is intended to be only used when
+ * {@see \PackageVersions\Versions} fails to be generated, which typically
+ * happens when running composer with `--no-scripts` flag)
+ */
+final class FallbackVersions
+{
+    private function __construct()
+    {
+    }
+
+    /**
+     * @param string $packageName
+     *
+     * @return string
+     *
+     * @throws \OutOfBoundsException if a version cannot be located
+     * @throws \UnexpectedValueException if the composer.lock file could not be located
+     */
+    public static function getVersion(string $packageName) : string
+    {
+        $versions = iterator_to_array(self::getVersions(self::getComposerLockPath()));
+
+        if (! array_key_exists($packageName, $versions)) {
+            throw new \OutOfBoundsException(
+                'Required package "' . $packageName . '" is not installed: cannot detect its version'
+            );
+        }
+
+        return $versions[$packageName];
+    }
+
+    /**
+     * @return string
+     *
+     * @throws \UnexpectedValueException
+     */
+    private static function getComposerLockPath() : string
+    {
+        // bold assumption, but there's not here to fix everyone's problems.
+        $checkedPaths = [__DIR__ . '/../../../../../composer.lock', __DIR__ . '/../../composer.lock'];
+
+        foreach ($checkedPaths as $path) {
+            if (file_exists($path)) {
+                return $path;
+            }
+        }
+
+        throw new \UnexpectedValueException(sprintf(
+            'PackageVersions could not locate your `composer.lock` location. This is assumed to be in %s. '
+            . 'If you customized your composer vendor directory and ran composer installation with --no-scripts, '
+            . 'then you are on your own, and we can\'t really help you. Fix your shit and cut the tooling some slack.',
+            json_encode($checkedPaths)
+        ));
+    }
+
+    private static function getVersions(string $composerLockFile) : \Generator
+    {
+        $lockData = json_decode(file_get_contents($composerLockFile), true);
```

Comment:
Yes, if the package is not installed (because of non-dev install) then the
version will still be retrievable

On 22 Jul 2016 18:12, "Christophe Coevoet" <notifications@github.com> wrote:

> In src/PackageVersions/FallbackVersions.php
> <https://github.com/Ocramius/PackageVersions/pull/33#discussion_r71904545>
> :
>
> > +            if (file_exists($path)) {
> > +                return $path;
> > +            }
> > +        }
> > +
> > +        throw new \UnexpectedValueException(sprintf(
> > +            'PackageVersions could not locate your `composer.lock` location. This is assumed to be in %s. '
> > +            . 'If you customized your composer vendor directory and ran composer installation with --no-scripts, '
> > +            . 'then you are on your own, and we can\'t really help you. Fix your shit and cut the tooling some slack.',
> > +            json_encode($checkedPaths)
> > +        ));
> > +    }
> > +
> > +    private static function getVersions(string $composerLockFile) : \Generator
> > +    {
> > +        $lockData = json_decode(file_get_contents($composerLockFile), true);
>
> Oh, so it is OK if the package is not there anymore ? This should be
> clearly documented then
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/Ocramius/PackageVersions/pull/33/files/0a1a91ea4d08baa225724845431bee542773fd1a#r71904545>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakNm1YtoJ-XgYCngwzUo3k-RhKxvcks5qYOv7gaJpZM4JSv1j>
> .
>